### PR TITLE
fixes issue #15: allow named sizes for named dimensions

### LINKED ISSUE #15: allow named sizes for named dimensions
(Thanks for this awesome work!)

I've often seen applications in which multiple dimensions of the same tensor will be the same size (which size is only known at runtime).  It is useful to be able to separately document that those sizes must be the same while acknowledging the differing purposes of these dimensions.

The following example demonstrates the idea---showing three related features that already work in torchtyping as well as a proposed feature:
```python
def func(feats: TensorType["b": ..., "annotator": 3, "word": "words", "feature"],
         predicates: TensorType["b": ..., "annotator": 3, "predicate": "words", "feature"],
         pred_arg_pairs: TensorType["b": ..., "annotator": 3, "predicate": "words", "argument": "words"]):
    # feats has shape (..., 3, words, features)
    # predicates has shape (..., 3, words, features)
    # pred_arg_pairs has shape (..., 3, words, words)
    # the ... b dimensions are checked to be of the same size.
```

Things that already work:
- dimensions that share names (like "feature") are enforced to share the same dimension size
- named dimensions (like "annotator") can specify a specific dimension size which is enforced
- named ellipses (like "batch") can be used to represent a fixed (but only known at runtime) set of dimensions and corresponding sizes [this is very close to what I want, but (1) I would prefer to not have the extra power of matching an unspecified number of dimensions and (2) as I understand it, ellipses only represent a single variable set of dimensions, but I want to be able to separately constrain multiple sets of dimensions to share respective sizes]

Proposed:
- named dimensions (like "token", "predicate", and "argument") should be able to declare a shared-but-unspecified dimension size given by name ("words" in this example)

Additionally, you would probably want to enforce that, if the specified "size name" matches the name of another dimension (like "word" in the following example), then the sizes of those dimensions should be the same:
```python
def func(feats: TensorType["b": ..., "annotator": 3, "word", "feature"],
         predicates: TensorType["b": ..., "annotator": 3, "predicate": "word", "feature"],
         pred_arg_pairs: TensorType["b": ..., "annotator": 3, "predicate": "word", "argument": "word"]):
    # feats has shape (..., 3, words, features)
    # predicates has shape (..., 3, words, features)
    # pred_arg_pairs has shape (..., 3, words, words)
    # the ... b dimensions are checked to be of the same size.
```

Thoughts?

### PR DESCRIPTION
**Pull request by teichert**
fixes #15, tests, and updates README

### CODE REVIEW

File: test/test_examples.py [hunk 1]
```diff
@@ -223,3 +224,37 @@ def func(x: TensorType[sparse_coo]):
...
+    with pytest.raises(TypeError):
+        func(ones(2, 1, 2, 4, 5), ones(2, 1, 2, 4, 5), ones(2, 1, 2, 4, 4))
+
+    # predicate and argument match word
+    with pytest.raises(TypeError):
+        func(ones(2, 1, 3, 3, 5), ones(2, 2, 3, 4, 5), ones(2, 1, 3, 4, 4))
+    with pytest.raises(TypeError):
+        func(ones(2, 1, 3, 4, 5), ones(2, 2, 3, 3, 5), ones(2, 1, 3, 4, 4))
+    with pytest.raises(TypeError):
+        func(ones(2, 1, 3, 4, 5), ones(2, 2, 3, 4, 5), ones(2, 1, 3, 3, 4))
```
Comment (patrick-kidger) [inline test/test_examples.py:260]: The middle argument for these last three checks has size `2, 2, 3, 4, 5` when it should presumably have `2, 1, 3, 4, 5` to match the `...`.

File: test/test_shape.py [hunk 2]
```diff
@@ -147,6 +148,93 @@ def _m1bm1_dim_checker(x: TensorType[-1, "b", -1]):
...
+        return torch.ones(x.shape[0])
+
+    func(torch.ones(3, names=["a"]))
+    with pytest.raises(TypeError):
+        func(torch.ones(3))
+    with pytest.raises(TypeError):
+        func(torch.ones(3, names=["b"]))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, names=["a"]))
+
```
Comment (patrick-kidger) [inline test/test_shape.py:236]: A test that `names=["x"]` fails would be good to have as well, given that that's the other name.

Or perhaps we should make it permissive and accept either name on a named dimension? No strong feelings either way, just that it should be checked + documented whichever way around we do it.

File: torchtyping/tensor_details.py [hunk 3]
```diff
@@ -97,6 +97,8 @@ def check(self, tensor: torch.Tensor) -> bool:
                 and self_name != tensor_name
             ):
                 return False
+            if isinstance(self_size, str):
+                return True
```
Comment (patrick-kidger) [inline torchtyping/tensor_details.py:101]: This looks wrong, as this breaks out of the loop early. Only `return False` should appear in the loop, and it's only if all of those checks pass that we then hit the `return True` at the end of the loop.

Instead I think this should be merged into the next check: `if (self_size not in (-1, tensor_size)) or isinstance(self_size, str): return False`

**Review by patrick-kidger — COMMENTED** [review]
For the most part, this LGTM. I've left a few comments. Besides this:

Presumably `None: str` pairs should also be allowed: corresponding to a named tensor without a name down that dimension, but for which we want to check the size according to the string.

Can you add a test that `other: str` pairs fail, for `other` not `str, None`, just to make sure we're parsing correctly?

+1 on the thorough tests you've already added, those are really good.

File: test/test_shape.py (hunk 2 above)
Comment (teichert) [reply test/test_shape.py:287]: good idea!  I do think that we should strictly check that the name mates the left, though.  Test added.

File: torchtyping/tensor_details.py (hunk 3 above)
Comment (teichert) [reply torchtyping/tensor_details.py]: you're right; I missed the loop! (I tried hard to find a test-case that would fail before fixing this, but because some things are checked twice, I didn't actually manage to find one).  I fixed this and added a test that would "almost" fail without the fix. :)

**Comment by patrick-kidger** [comment]
Thanks for the changes.

I actually wouldn't worry about that last one. For the time being any string is a valid way of referring to a dimension, and this then becomes an exception to that rule. And once #2 happens, I don't see a reason to avoid functions in the other named dimensoin.

**Comment by teichert** [comment]
I went ahead and made the change so that only legal identifiers are allowed in size strings.  (That will be easy to remove when more complex expressions are allowed, but I think it is important so that no one thinks that "a": "x+y" is enforcing something that it isn't.) Would you prefer that I revert?
